### PR TITLE
[Parser][NFC] Create a public wat-lexer.h header

### PR DESCRIPTION
**Pull request by tlively**
wat-parser-internal.h was already quite large after implementing just the lexer,
so it made sense to rename it to be lexer-specific and start a new file for the
higher-level parser. Also make it a proper .cpp file and split the testable
interface out into wat-lexer.h.

### CODE REVIEW

**Comment by tlively** [comment]
Current dependencies on/for this PR:
* main
  * **PR #4693** <a href="https://app.graphite.dev/github/pr/WebAssembly/binaryen/4693" target="_blank">Graphite</a> 
    * **PR #4694** <a href="https://app.graphite.dev/github/pr/WebAssembly/binaryen/4694" target="_blank">Graphite</a> 
      * **PR #4695** <a href="https://app.graphite.dev/github/pr/WebAssembly/binaryen/4695" target="_blank">Graphite</a>  👈
        * **PR #4696** <a href="https://app.graphite.dev/github/pr/WebAssembly/binaryen/4696" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/WebAssembly/binaryen/4695?utm_source=stack-comment).

**Review by aheejin — COMMENTED** [review]
You said you named the previous header `wat-parser-internal.h` because this was not meant to be used from outside, and you were gonna create a new header with public API. Now these headers don't have `internal` in the names, are they meant to be public? But this has the same contents as the previous `wat-parser-internal.h`.

File: test/gtest/wat-lexer.cpp
Comment (tlively) [reply test/gtest/wat-lexer.cpp:1]: Looks like it does now 👍

**Comment by tlively** [comment]
If there were a good way to make a hierarchy of headers, then I would make `wat-lexer.h` a private header within the parser module so that it could be used from a future `wat-parser.cpp` and from nowhere else. Unfortunately there isn't a great way to do that in C++. If we're going to split the lexer's interface declarations from its implementation (which I think makes sense given how large the implementation is), then I don't see any benefit in marking the header "internal" anymore.

I doubt any other code in binaryen will want to use the lexer directly, but if it does, that would be ok after this change.

**Review by kripken — APPROVED** [review]

**Review by aheejin — APPROVED** [review]

**Comment by tlively** [comment]
**Graphite Merge Job**

Current status: 🔄 Waiting for checks to pass

This pull request is currently waiting for CI and mergeability checks to pass before being merged.

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/WebAssembly/binaryen/4695?utm_source=stack-comment).


Job Reference: UksnMFlJsREOdJP1qpqD

**Pull request merged**
